hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/multipart/MultipartDownloader.java

```diff
@@ -0,0 +1,103 @@
+package org.apache.hadoop.fs.s3a.multipart;
+
+import com.google.common.collect.Lists;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.util.List;
+
+public final class MultipartDownloader {
+    private static final Log LOG = LogFactory.getLog(MultipartDownloader.class);
+
+    private final long partSize;
+    private final ListeningExecutorService downloadExecutorService;
+    private final PartDownloader partDownloader;
+    private final long chunkSize;
+    private final long bufferSize;
+
+    public MultipartDownloader(long partSize, ListeningExecutorService downloadExecutorService, PartDownloader partDownloader, long chunkSize, long bufferSize) {
+        this.partSize = partSize;
+        this.downloadExecutorService = downloadExecutorService;
+        this.partDownloader = partDownloader;
+        this.chunkSize = chunkSize;
```

Comment:
Do we need `Preconditions.checkArgument(chunkSize < Integer.MAX_VALUE - 8)` (as you typically can’t allocate an array larger than that per
http://hg.openjdk.java.net/jdk8/jdk8/jdk/rev/ec45423a4700#l4.19)